File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/LanguageConfiguration.java
package com.qiplat.sweeteditor;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Language configuration describing metadata such as brackets, comments, and indentation for a specific programming language.
 * When set to EditorCore, brackets will be automatically synced to Core layer's setBracketPairs.
 */
public class LanguageConfiguration {

    public static final int DEFAULT_TAB_SIZE = 4;

    /** Bracket pair */
    public static class BracketPair {
        @NonNull public final String open;
        @NonNull public final String close;

        public BracketPair(@NonNull String open, @NonNull String close) {
            this.open = open;
            this.close = close;
        }
    }

    @NonNull private final String languageId;
    @Nullable private final List<BracketPair> brackets;
    @Nullable private final List<BracketPair> autoClosingPairs;
    private final int tabSize;
    private final boolean insertSpaces;

    private LanguageConfiguration(Builder builder) {
        this.languageId = builder.languageId;
        this.brackets = builder.brackets == null ? null : Collections.unmodifiableList(new ArrayList<>(builder.brackets));
        this.autoClosingPairs = builder.autoClosingPairs == null ? null : Collections.unmodifiableList(new ArrayList<>(builder.autoClosingPairs));
        this.tabSize = builder.tabSize;
        this.insertSpaces = builder.insertSpaces;
    }

    @NonNull public String getLanguageId() { return languageId; }
    @Nullable public List<BracketPair> getBrackets() { return brackets; }
    @Nullable public List<BracketPair> getAutoClosingPairs() { return autoClosingPairs; }
    public int getTabSize() { return tabSize; }
    public boolean getInsertSpaces() { return insertSpaces; }

    public static class Builder {
        @NonNull private final String languageId;
        private List<BracketPair> brackets;
        private List<BracketPair> autoClosingPairs;
        private int tabSize = DEFAULT_TAB_SIZE;
        private boolean insertSpaces;

        public Builder(@NonNull String languageId) {
            this.languageId = languageId;
        }

        public Builder addBracket(@NonNull String open, @NonNull String close) {
            if (brackets == null) brackets = new ArrayList<>();
            brackets.add(new BracketPair(open, close));
            return this;
        }

        public Builder addAutoClosingPair(@NonNull String open, @NonNull String close) {
            if (autoClosingPairs == null) autoClosingPairs = new ArrayList<>();
            autoClosingPairs.add(new BracketPair(open, close));
            return this;
        }

        public Builder setTabSize(int tabSize) {
            this.tabSize = tabSize;
            return this;
        }

        public Builder setInsertSpaces(boolean insertSpaces) {
            this.insertSpaces = insertSpaces;
            return this;
        }

        public LanguageConfiguration build() {
            return new LanguageConfiguration(this);
        }
    }
}

File: platform/Apple/Sources/SweetEditorCoreInternal/api/LanguageConfiguration.swift
import Foundation

// MARK: - LanguageConfiguration

/// Language configuration describing language-specific metadata such as brackets, comments, and indentation.
/// When set on SweetEditorCore, bracket pairs are automatically synced to Core via `setBracketPairs`.
public struct LanguageConfiguration {

    public static let defaultTabSize = 4

    /// Bracket pair.
    public struct BracketPair {
        public let open: String
        public let close: String

        public init(open: String, close: String) {
            self.open = open
            self.close = close
        }
    }

    /// Block comment pair.
    public struct BlockComment {
        public let open: String
        public let close: String

        public init(open: String, close: String) {
            self.open = open
            self.close = close
        }
    }

    /// Language identifier (for example: "swift", "java", "cpp").
    public let languageId: String

    /// Bracket pair list (synced to Core `setBracketPairs`). nil means not configured.
    public let brackets: [BracketPair]?

    /// Auto-closing pair list (used by platform-side auto-closing logic). nil means not configured.
    public let autoClosingPairs: [BracketPair]?

    /// Line-comment prefix (for example: "//").
    public let lineComment: String?

    /// Block-comment pair (for example: ("/*", "*/")).
    public let blockComment: BlockComment?

    /// Tab width (optional; if omitted, editor default is used).
    public let tabSize: Int?

    /// Whether to use spaces instead of tabs (optional).
    public let insertSpaces: Bool?

    public init(languageId: String,
                brackets: [BracketPair]? = nil,
                autoClosingPairs: [BracketPair]? = nil,
                lineComment: String? = nil,
                blockComment: BlockComment? = nil,
                tabSize: Int? = nil,
                insertSpaces: Bool? = nil) {
        self.languageId = languageId
        self.brackets = brackets
        self.autoClosingPairs = autoClosingPairs
        self.lineComment = lineComment
        self.blockComment = blockComment
        self.tabSize = tabSize
        self.insertSpaces = insertSpaces
    }
}
